include/interpolation_xz.hxx

```diff
@@ -0,0 +1,252 @@
+
+  Field3D t_x, t_z;
+
+public:
+  XZLagrange4pt(Mesh *mesh = nullptr, Options* opt = nullptr)
+      : XZLagrange4pt(0, mesh, opt) {}
+  XZLagrange4pt(int y_offset = 0, Mesh *mesh = nullptr, Options* opt = nullptr);
+  XZLagrange4pt(const BoutMask &mask, int y_offset = 0, Mesh *mesh = nullptr,
+                Options* opt = nullptr)
+      : XZLagrange4pt(y_offset, mesh, opt) {
+    skip_mask = mask;
+  }
+
+  void calcWeights(const Field3D &delta_x, const Field3D &delta_z) override;
+  void calcWeights(const Field3D &delta_x, const Field3D &delta_z,
+                   const BoutMask &mask) override;
+
+  // Use precalculated weights
+  Field3D interpolate(const Field3D &f) const override;
+  // Calculate weights and interpolate
+  Field3D interpolate(const Field3D &f, const Field3D &delta_x,
+                      const Field3D &delta_z) override;
+  Field3D interpolate(const Field3D &f, const Field3D &delta_x, const Field3D &delta_z,
+                      const BoutMask &mask) override;
+  BoutReal lagrange_4pt(BoutReal v2m, BoutReal vm, BoutReal vp, BoutReal v2p,
+                        BoutReal offset) const;
+  BoutReal lagrange_4pt(const BoutReal v[], BoutReal offset) const;
+};
+
+class XZBilinear : public XZInterpolation {
+  Tensor<int> i_corner; // x-index of bottom-left grid point
+  Tensor<int> k_corner; // z-index of bottom-left grid point
+
+  Field3D w0, w1, w2, w3;
+
+public:
+  XZBilinear(Mesh *mesh = nullptr, Options* opt = nullptr) : XZBilinear(0, mesh, opt) {}
+  XZBilinear(int y_offset = 0, Mesh *mesh = nullptr, Options* opt = nullptr);
+  XZBilinear(const BoutMask &mask, int y_offset = 0, Mesh *mesh = nullptr,
+             Options* opt = nullptr)
+      : XZBilinear(y_offset, mesh, opt) {
+    skip_mask = mask;
+  }
+
+  void calcWeights(const Field3D &delta_x, const Field3D &delta_z) override;
+  void calcWeights(const Field3D &delta_x, const Field3D &delta_z,
+                   const BoutMask &mask) override;
+
+  // Use precalculated weights
+  Field3D interpolate(const Field3D &f) const override;
+  // Calculate weights and interpolate
+  Field3D interpolate(const Field3D &f, const Field3D &delta_x,
+                      const Field3D &delta_z) override;
+  Field3D interpolate(const Field3D &f, const Field3D &delta_x, const Field3D &delta_z,
+                      const BoutMask &mask) override;
+};
+
+class XZInterpolationFactory
+    : public Factory<XZInterpolation, XZInterpolationFactory,
+                     std::function<std::unique_ptr<XZInterpolation>(Mesh*, Options*)>> {
+public:
+  static constexpr auto type_name = "XZInterpolation";
+  static constexpr auto section_name = "xzinterpolation";
+  static constexpr auto option_name = "type";
+  static constexpr auto default_type = "hermitespline";
+
+  using Factory::create;
+  ReturnType create(Mesh* mesh = nullptr, Options* opt = nullptr) {
+    return Factory::create(getType(nullptr), mesh, opt);
```

Comment:
If this were `Factory::create(getType(opt), mesh, opt)`, then the call in `fci.cxx` could avoid the repeated argument.

Reply:
Second argument is not there any more, because of removing `Options` arguments from `XZInterpolation` constructors.